File: src/game/controller/ControllerManager.java

package game.controller;

import game.Bomberman;
import game.engine.PeriodicTask;
import game.events.Observable2;
import game.utils.Utility;

import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.*;

import static java.util.Map.entry;

/**
 This class represents an Observable object that observes key events and notifies its observers of the
 corresponding command that should be executed based on the key that was pressed.
 */
public class ControllerManager extends Observable2 implements KeyListener {
    private static ControllerManager instance;
    private static final int KEY_W = KeyEvent.VK_W;
    private static final int KEY_A = KeyEvent.VK_A;
    private static final int KEY_S = KeyEvent.VK_S;
    private static final int KEY_D = KeyEvent.VK_D;
    private static final int KEY_SPACE = KeyEvent.VK_SPACE;
    private static final int KEY_ESC = KeyEvent.VK_ESCAPE;
    private static int KEY_DELAY_MS = setDefaultCommandDelay();
    public Set<Command> commandQueue = new HashSet<>();

    // Key-Command mapping
    private static final Map<Integer, Command> keyAssignment = Map.ofEntries(
            entry(KEY_W, Command.MOVE_UP),
            entry(KEY_A, Command.MOVE_LEFT),
            entry(KEY_S, Command.MOVE_DOWN),
            entry(KEY_D, Command.MOVE_RIGHT),
            entry(KEY_SPACE, Command.PLACE_BOMB),
            entry(KEY_ESC, Command.PAUSE)
    );

    // Stores the time of the last key event for each command
    private final Map<Command, Long> commandEventsTime = new HashMap<>();
    private PeriodicTask task;

    public ControllerManager(){
        instance = this;
        setupTask();
    }

    public void onKeyPressed(Command action){
        // Ignore the event if the time elapsed since the last event is less than KEY_DELAY_MS
        if(Utility.timePassed( commandEventsTime.getOrDefault(action, 0L)) < KEY_DELAY_MS) return;

        if(action != null) {
            commandEventsTime.put(action, System.currentTimeMillis());
            commandQueue.add(action);
        }

        resume();

    }

    /**
     * Handles key pressed events and notifies observers with the corresponding command that should be executed.
     *
     * @param e the KeyEvent object that contains the information of the key that was pressed.
     */
    @Override
    public void keyPressed(KeyEvent e) {
        Command action = keyAssignment.get(e.getKeyCode());
        onKeyPressed(action);
        // if a button is pressed, mouse movement gets interrupted
        Bomberman.getMatch().getMouseControllerManager().stopPeriodicTask();
    }

    @Override
    public void keyReleased(KeyEvent e) {
        Command action = keyAssignment.get(e.getKeyCode());
        onKeyReleased(action);

    }
    public void onKeyReleased(Command action){
        commandQueue.remove(action);
        if(commandQueue.isEmpty()) stop();
    }

    private void setupTask() {
        task = new PeriodicTask(() -> {
            for (Command command : new HashSet<>(commandQueue)) {
                notifyObservers(command);
            }
        }, KEY_DELAY_MS);

        task.start();
    }

    private void resume() {
        try{
            if(task != null) {
                task.resume();
            }
        }catch (Exception ignored){}
    }

    private void stop() {
        try{
            if(task != null) task.stop();
        }catch (Exception ignored){}
    }

    public boolean isCommandPressed(Command c) {
        return commandQueue.contains(c);
    }

    private void updateDelay() {
        instance.task.setDelay(KEY_DELAY_MS);
    }

    public static int decreaseCommandDelay() {
        KEY_DELAY_MS = 15;
        if (instance != null) {
            instance.updateDelay();
        }
        return KEY_DELAY_MS;
    }

    public static int setDefaultCommandDelay(){
        KEY_DELAY_MS = 30;
        if (instance != null) {
            instance.updateDelay();
        }
        return KEY_DELAY_MS;
    }

    @Override public void keyTyped(KeyEvent e) {}
}

File: src/game/input/ControllerManager.kt
package game.input

import game.JBomb
import game.data.data.DataInputOutput
import game.domain.tasks.observer.Observable2
import game.utils.Utility.timePassed
import game.utils.time.now
import java.awt.event.KeyEvent
import java.awt.event.KeyListener

/**
 * This class represents an Observable object that observes key events and notifies its observers of the
 * corresponding command that should be executed based on the key that was pressed.
 */
class ControllerManager : Observable2(), KeyListener {
    private val commandQueue: HashSet<Command> = hashSetOf()

    // Stores the time of the last key event for each command
    private val commandEventsTime: MutableMap<Command?, Long> = mutableMapOf()
    // Key-Command mapping
    private var keyAssignment: Map<Int, Command>? = null

    init {
        instance = this
        // If illegal keys are found, reset the key map;
        try {
            setKeyMap()
        } catch (e: IllegalArgumentException) {
            DataInputOutput.getInstance().resetKeys()
            DataInputOutput.getInstance().updateStoredPlayerData()
            setKeyMap()
            e.printStackTrace()
        }
    }


    private fun setKeyMap() {
        val dataInputOutput = DataInputOutput.getInstance()

        keyAssignment = mapOf(
                dataInputOutput.forwardKey to Command.MOVE_UP,
                dataInputOutput.leftKey to Command.MOVE_LEFT,
                dataInputOutput.backKey to Command.MOVE_DOWN,
                dataInputOutput.rightKey to Command.MOVE_RIGHT,
                dataInputOutput.bombKey to Command.ATTACK,
                dataInputOutput.interactKey to Command.INTERACT,
                KEY_ESC to Command.PAUSE
        )
    }

    private fun pressKey(action: Command) {
        // if a button is pressed, mouse movement gets interrupted
        if (action.isMovementKey()) {
            JBomb.match.mouseControllerManager.stopMovementTask()
        }

        onKeyPressed(action)
    }

    fun onKeyPressed(action: Command) {
        // Ignore the event if the time elapsed since the last event is less than KEY_DELAY_MS
        commandEventsTime[action] = now()
        notifyObservers(ObserverParam(ObserverParamIdentifier.INPUT_COMMAND, action))
    }

    /**
     * Handles key pressed events and notifies observers with the corresponding command that should be executed.
     *
     * @param e the KeyEvent object that contains the information of the key that was pressed.
     */
    override fun keyPressed(e: KeyEvent) {
        val action = keyAssignment?.get(e.keyCode) ?: return

        if (timePassed(commandEventsTime.getOrDefault(action, 0L)) < KEY_DELAY_MS)
            return

        pressKey(action)
    }

    override fun keyReleased(e: KeyEvent) {
        val action = keyAssignment?.get(e.keyCode) ?: return
        onKeyReleased(action)
    }

    fun onKeyReleased(action: Command?) {
        notifyObservers(ObserverParam(ObserverParamIdentifier.DELETE_COMMAND, action))
        commandQueue.remove(action)
    }

    fun isCommandPressed(c: Command): Boolean {
        return commandQueue.contains(c)
    }

    override fun keyTyped(e: KeyEvent) {}

    companion object {
        private const val KEY_ESC = KeyEvent.VK_ESCAPE
        private var instance: ControllerManager? = null
        private var KEY_DELAY_MS = 30
    }
}